crates/publish/src/oci/client.rs

```diff
@@ -209,28 +214,86 @@ impl Client {
         ))
     }
 
+    /// Save a credential set containing the registry username and password.
+    pub async fn login(
+        server: impl AsRef<str>,
+        username: impl AsRef<str>,
+        password: impl AsRef<str>,
+    ) -> Result<()> {
+        // We want to allow a user to login to both https://ghcr.io and ghcr.io.
+        let server = match server.as_ref().parse::<Url>() {
+            Ok(url) => url.host_str().unwrap_or(server.as_ref()).to_string(),
+            Err(_) => server.as_ref().to_string(),
+        };
+
+        // First, test the credentials. If a user accidentally enters a wrong credential set, this
+        // can catch the issue early rather than the user having to figure it out on their own.
+        Self::test_credentials(&server, &username, &password)
+            .await
+            .context("cannot check credentials")?;
+
+        // Save an encoded representation of the credential set in the local configuration file.
+        let mut auth = AuthConfig::load_default().await?;
+        auth.insert(server, username, password)?;
+        auth.save_default().await
+    }
+
+    /// Test the credentials by attempting to send an authenticated request to the registry.
+    async fn test_credentials(
+        server: impl AsRef<str>,
+        username: impl AsRef<str>,
+        password: impl AsRef<str>,
+    ) -> Result<()> {
+        let client = dkregistry::v2::Client::configure()
+            .registry(server.as_ref())
+            .insecure_registry(false)
+            .username(Some(username.as_ref().into()))
```

Comment:
This suggests that your args should be `impl Into<String>`.

Reply:
I wouldn't optimize the arguments for that call, since it is most likely going away in a future version.